src/os/bluestore/BlueStore.cc

```diff
@@ -3701,10 +3701,11 @@ void BlueStore::Onode::put() {
         c->onode_map._remove(oid);
       }
     }
+    n = nref;
+    --put_nref;
     ocs->lock.unlock();
   }
-  auto pn = --put_nref;
-  if (nref == 0 && pn == 0) {
+  if (n == 0 && put_nref == 0) {
```

Comment:
First of all - it's not mandatory that onode is cached and hence potentially two threads might own onode independently with no references from c->onode_map.
So we might have two threads owning onode with nref == 2. Finally both threads call put()....
Case 1:
Thread A makes nref ==1 and goes through if(n == 1) block but before it reaches --put_nref thread B falls through and deletes onode... At this point thread A operates on a released onode.

Case 2: 
Thread A makes nref == 1 and reaches n = nref. At this point thread B makes nref == 0 and falls through to put() return - with no onode release due to put_nref != 0. Then thread A continues and bypasses delete due to n == 1 as well. Hence Onode is leaking...

Generally my idea behind put_ref increment/decrement is that it has to "wrap"  other manipulations on onode within put(). So its increment has to be the first op in put() and decrement to be the last one before the delete.  I haven't achieved that completely but looks like you're moving even further from the original idea...

So we don't have good enough fix for now :(

Reply:
Thank you very much for your review. 
I did really assume caching onode is mandatory -- an Onode had to be added to the cache before it could be referenced. 
Could you please give me some examples that the Onode might be referenced without adding to cache ? deep-scrub ?

Reply:
@taodd - please take another look at https://github.com/ceph/ceph/pull/47702
I reworked my patch and it looks better now ;)

Reply:
such as two onode cached in two threads, but the onode is trimed.